Support multidimensional tensors in data viewer

joyceerhl opened the pull request:
For https://github.com/microsoft/vscode-jupyter/issues/298

Behavior changes:
1. All values will now be truncated past 10 characters per @jmew
2. Higher dimensional data will be represented as flattened strings with the ability to view the full value
3. Full view is readonly


Note, I wasn't able to bind esc/enter to exit and enter edit mode with `autoEdit: false`--there's a weird bug where after exiting edit mode, focus gets lost. If we allow auto edits as soon as a cell receives focus (current behavior, see GIF below for illustration), this preserves the custom cell navigation behavior implemented in our keydown handlers. I'll log that as a separate bug since I think investigating that is going to be somewhat involved.

![tabbing](https://user-images.githubusercontent.com/30305945/105916486-d32ca500-5fe5-11eb-9f02-5ec23c8744ba.gif)


Multidimensional tensors:
![image](https://user-images.githubusercontent.com/30305945/105787925-4a096580-5f34-11eb-9c0b-7410c1848cb9.png)

Ragged ndarrays:
![image](https://user-images.githubusercontent.com/30305945/105786568-30671e80-5f32-11eb-9cf2-4ef0aa61b1c8.png)

Tooltip contains the full value:
![image](https://user-images.githubusercontent.com/30305945/105786554-29401080-5f32-11eb-85c1-5e17a5a696f3.png)


<!--
  If an item below does not apply to you, then go ahead and check it off as "done" and strikethrough the text, e.g.:
    - [x] ~Has unit tests & system/integration tests~
-->

-   [x] Pull request represents a single change (i.e. not fixing disparate/unrelated things in a single PR).
-   [x] Title summarizes what is changing.
-   [x] Has a [news entry](https://github.com/Microsoft/vscode-jupyter/tree/main/news) file (remember to thank yourself!).
-   [x] Appropriate comments and documentation strings in the code.
-   [x] Has sufficient logging.
-   [x] ~Has telemetry for enhancements.~
-   [x] Unit tests & system/integration tests are added/updated.
-   [x] [Test plan](https://github.com/Microsoft/vscode-jupyter/blob/main/.github/test_plan.md) is updated as appropriate.
-   [x] ~[`package-lock.json`](https://github.com/Microsoft/vscode-jupyter/blob/main/package-lock.json) has been regenerated by running `npm install` (if dependencies have changed).~

rchiodo commented on pythonFiles/vscode_datascience_helpers/dataframes/vscodeDataFrame.py:
```diff
@@ -1,11 +1,57 @@
+            for j in range(len(temp[i])):
+                element = temp[i][j]
+                if isinstance(element, _VSCODE_np.ndarray):
+                    stringified = _VSCODE_np.array2string(element, separator=", ")
```
I thought you found this was super slow and you were going to do this in the UI?

joyceerhl replied:
No, I found that lazy fetching details of individual cells was too slow (2 seconds for long values). Here I'm requesting the full string and then doing all the truncation in the UI.

rchiodo replied:
Oh so you're concatenating the value here (instead of sending the full json).

rchiodo replied:
Sorry I though the slowdown in the UI meant you were going to do the concat there as well.

joyceerhl replied:
Bit confused. By 'concatenating' do you mean transforming the data from an array to a string?

Just to be sure we're on the same page, the intent in this bit of code is to generate the full string representation of all data past 2 dimensions, without any truncation, over to the UI. Then in the cellFormatter I'm displaying all values with a trailing ellipsis, so we only render the first 10 characters of any value. We display the full untruncated value later.

In a separate PR I'll also implement pagination like we talked about, which should minimize the performance impact of sending untruncated strings to the UI.

rchiodo replied:
Yes I meant transforming the data from an array to a string. Theoretically the json for the array isn't much more than the json for the string so I thought you were just going to send the array itself over. But maybe that is significantly slower.

And I believe you'll be throwing this code away when you implement slicing?

joyceerhl replied:
> I thought you were just going to send the array itself over

I think this doesn't really work for >3D data. Python objects know how to represent themselves as strings in Python code, but we can't get the string representation out of them in TypeScript. At minimum we'd have to recursively convert Python objects like Tensors or numpy ndarrays to arrays, and then I'm not sure how we'd reconstitute nested arrays.

> you'll be throwing this code away when you implement slicing

Unfortunately not, because the slice data view as currently spec-ed only applies to 3D data. We've talked to Jill about reworking the design, but I believe the current plan is to maybe extend it to work for 2D or 3D data. 

If we made the slice data view work for arbitrary dimensions and selected default dimensions for higher dimensional arrays, then yes we'd be able to deprecate this code. Jeffrey and I have discussed whether the latter would be confusing for users (e.g. they might think the data is 2D because we default them to a flat view with a preselected slice). So the current plan is to show the flattened view (implemented in this PR) by default, and then also provide a way to turn on slicing mode.

joyceerhl commented on src/datascience-ui/data-explorer/cellFormatter.tsx:
```diff
@@ -74,3 +72,12 @@ export function cellFormatterFunc(
+    if (value.length <= 10) {
+        return value;
+    } else {
+        return value.slice(0, 7).concat('...');
```
This part here is where we do the truncation now @rchiodo

IanMatthewHuff commented on pythonFiles/vscode_datascience_helpers/dataframes/vscodeDataFrame.py:
```diff
@@ -1,11 +1,57 @@
+    del temp
+
+    # Restore the user's printoptions
+    _VSCODE_np.set_printoptions(threshold=current_options["threshold"])
```
Should we possibly have everything in this function in a try and then restore this in a finally? Just worried about not restoring users print options.

joyceerhl replied:
Ah you're right. I was trying to minimize my use of expansive try/catch blocks (got feedback from Eric on that in my TypeScript code 😅)

FWIW this Python code all needs to move to the Python extension. But given the hard deadline on this work for March I'm planning to get to that refactor afterwards, since that involves designing an API for providing debug info for variable views as well as the data viewer.

IanMatthewHuff replied:
Interesting, was Eric's feedback python specific or just in general? I would assume that an exception here would be a true exceptional case (i.e. not something that we know will happen from time to time) so in the non-exception thrown case I don't think this would be a perf issue.

joyceerhl replied:
His feedback was general and in the context of regular TypeScript code, so I don't know that it applies here. But I wrote this particular catch block accounts with something that will happen from time to time in mind. I think there's a way to handle that error without incurring the exception, but we would still want to avoid bombing out on an unexpected error here.

IanMatthewHuff reviewed (approved):
Looks legit to me. Though I have to confess that I'm pretty weak with Python :D. Just one comment regarding making sure that we restore user's print options.

joyceerhl commented:
Just updated this PR to allow users to view and scroll the expanded view of the cell using left/right arrows. I added a GIF to the issue description if y'all want to take a look @IanMatthewHuff and @rchiodo.
